frb/halos/hmf.py

```diff
@@ -0,0 +1,389 @@
+
+    Note that the fraction of DM associated with these halos
+    will be scaled down by an additional factor of f_diffuse
+
+    Requires Aemulus HMF to be installed
+
+    Args:
+        zvals: ndarray
+        Mlow: float
+          In h^-1 units already so this will be applied for the halo mass function
+        Mhigh: float
+          In h^-1 units already
+        rmax: float
+          Extent of the halo in units of rvir
+          WARNING: This calculation assumes a single concentration for all halos
+
+    Returns:
+        ratios: ndarray
+          rho_halo / rho_m
+    """
+
+    M = np.logspace(np.log10(Mlow*cosmo.h), np.log10(Mhigh*cosmo.h), num=1000)
+    lM = np.log(M)
+
+    ratios = []
+    for z in zvals:
+        # Setup
+        #dndlM = np.array([hmfe.dndlnM(Mi, a)[0] for Mi in M])
+        dndlM = M*hmfe.dndM(M, z)
+        M_spl = IUS(lM, M * dndlM)
+
+        # Integrate
+        rho_tot = M_spl.integral(np.log(Mlow*cosmo.h), np.log(Mhigh*cosmo.h)) * units.M_sun / units.Mpc ** 3
+        # Cosmology
+        rho_M = cosmo.critical_density(z) * cosmo.Om(z)/(1+z)**3  # Tinker calculations are all mass
+        ratio = (rho_tot*cosmo.h**2 / rho_M).decompose()
+        #
+        ratios.append(ratio)
+    ratios = np.array(ratios)
+    # Boost halos if extend beyond rvir (homologous in mass, but constant concentration is an approx)
+    if rmax != 1.:
+        #from pyigm.cgm.models import ModifiedNFW
+        c = 7.7
+        nfw = ModifiedNFW(c=c)
+        M_ratio = nfw.fy_dm(rmax * nfw.c) / nfw.fy_dm(nfw.c)
+        ratios *= M_ratio
+    # Return
+    return np.array(ratios)
+
+
+def halo_incidence(Mlow, zFRB, radius=None, hmfe=None, Mhigh=1e16, nsample=20,
+                   cumul=False):
+    """
+    Calculate the (approximate) average number of intersections to halos of a
+    given minimum mass to a given zFRB.
+
+    Requires Aemulus HMF to be installed
+
+    Args:
+        Mlow: float
+          Mass of minimum halo in Solar masses
+          The code deals with h^-1 factors so that you do not
+          The minimum value is 2e10
+        zFRB: float
+          Redshift of the FRB
+        radius: Quantity, optional
+          The calculation will specify this radius as rvir derived from
+           Mlow unless this is specified. And this rvir *will* vary with redshift
+        hmfe (hmf.hmf_emulator, optional): Halo mass function emulator from Aeumulus
+        Mhigh: float, optional
+          Mass of maximum halo in Solar masses
+        nsammple: int, optional
+          Number of samplings in redshift
+          20 should be enough
+        cumul: bool, optional
+          Return the cumulative quantities instead
+
+    Returns:
+        If cumul is False
+        Navg: float
+          Number of average intersections
+        elif cumul is True
+        zeval: ndarray
+        Ncumul: ndarray
+    """
+    # Mlow limit
+    if Mlow < 2e10:
+        warnings.warn("Calculations are limited to Mlow > 2e10")
```

Comment:
Why not just raise an error instead?

Reply:
I forget..  Must have broken something I didnt want broken.
But let me raise it instead..